controllers/virtualmachinesnapshot/virtualmachinesnapshot_controller.go

```diff
@@ -0,0 +1,170 @@
+// Reconciler reconciles a VirtualMachineSnapShot object.
+type Reconciler struct {
+	client.Client
+	Context  context.Context
+	Logger   logr.Logger
+	Recorder record.Recorder
+}
+
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachinesnapshots,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachinesnapshots/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachines,verbs=get;list;watch;
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachines/status,verbs=get;update;patch
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx = pkgcfg.JoinContext(ctx, r.Context)
+
+	vmSnapShot := &vmopv1.VirtualMachineSnapshot{}
+	if err := r.Get(ctx, req.NamespacedName, vmSnapShot); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	vmSnapShotCtx := &pkgctx.VirtualMachineSnapshotContext{
+		Context:                ctx,
+		Logger:                 ctrl.Log.WithName("VirtualMachineSnapShot").WithValues("name", req.Name),
+		VirtualMachineSnapshot: vmSnapShot,
+	}
+
+	patchHelper, err := patch.NewHelper(vmSnapShot, r.Client)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to init patch helper for %s: %w", vmSnapShotCtx, err)
+	}
+	defer func() {
+		vmSnapShotCtx.Logger.Info("Patching VirtualMachineSnapShot", "snap", vmSnapShot)
+		if err := patchHelper.Patch(ctx, vmSnapShot); err != nil {
+			if reterr == nil {
+				reterr = err
+			}
+			vmSnapShotCtx.Logger.Error(err, "patch failed")
+		}
+	}()
+
+	if !vmSnapShot.DeletionTimestamp.IsZero() {
+		// Noop.
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ReconcileNormal(vmSnapShotCtx); err != nil {
+		vmSnapShotCtx.Logger.Error(err, "Failed to reconcile VirtualMachineSnapShot")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineSnapshotContext) error {
+	ctx.Logger.Info("Reconciling VirtualMachineSnapshot")
+	vmSnapShot := ctx.VirtualMachineSnapshot
+
+	// return early if phase was already set; nothing to do
+	if conditions := ctx.VirtualMachineSnapshot.Status.Conditions; len(conditions) != 0 {
+		for _, c := range conditions {
+			if c.Type == vmopv1.VirtualMachineSnapshotReadyCondition {
+				return nil
+			}
+		}
+	}
+
+	ctx.Logger.Info("Fetching VirtualMachine from snapshot object", "vmSnapshot", vmSnapShot.Name)
+	vm := &vmopv1.VirtualMachine{}
+	objKey := client.ObjectKey{Name: vmSnapShot.Spec.VMRef.Name, Namespace: vmSnapShot.Namespace}
+	err := r.Get(ctx, objKey, vm)
+	if err != nil {
+		ctx.Logger.Error(err, "failed to get VirtualMachine", "vm", objKey)
+		return err
+	}
+	ctx.VM = vm
+
+	if vm.Status.UniqueID == "" {
+		err = errors.New("VM hasn't been created and has no uniqueID")
+		return err
+	}
+
+	objRef := &vmopv1common.LocalObjectRef{
+		APIVersion: vmSnapShot.APIVersion,
```

Comment:
With the entire local object ref being used for the DeepEquals, is the desired behavior really to take another snapshot when we rev our API version?

Reply:
makes sense. only checking the Name now